File: app/src/main/java/com/example/sasham/testproject/themes/ThemesListingFragment.java
package com.example.sasham.testproject.themes;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.example.sasham.testproject.BaseActivity;
import com.example.sasham.testproject.Constants;
import com.example.sasham.testproject.R;
import com.example.sasham.testproject.model.Theme;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import dagger.android.support.AndroidSupportInjection;

public class ThemesListingFragment extends Fragment implements ThemesListingView, BaseActivity.OnConnectionListener, SwipeRefreshLayout.OnRefreshListener {

    @Inject
    public ThemesListingPresenter presenter;

    private Callback callback;
    private List<Theme> themes = new ArrayList<>();

    @BindView(R.id.themes_listing_recycler_view)
    public RecyclerView themesRecyclerView;

    @BindView(R.id.themes_progress)
    public ProgressBar themesProgress;

    @BindView(R.id.themes_empty_view)
    public TextView themesEmptyView;

    @BindView(R.id.swipe_layout)
    SwipeRefreshLayout swipeRefreshLayout;

    private ThemesListingAdapter themesListingAdapter;
    private Unbinder unbinder;
    private BaseActivity baseActivity;
    private boolean isConnected = false;


    public ThemesListingFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof Callback) {
            callback = (Callback) context;
        }
        if (context instanceof BaseActivity) {
            baseActivity = (BaseActivity) context;
        }

    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        AndroidSupportInjection.inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View root = inflater.inflate(R.layout.fragment_themes_listing, container, false);

        //Init views
        unbinder = ButterKnife.bind(this, root);

        themesListingAdapter = new ThemesListingAdapter(themes, this);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        themesRecyclerView.setAdapter(themesListingAdapter);
        themesRecyclerView.setLayoutManager(layoutManager);

        //If we scrolled to end list then load more data
        themesRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);

                if (mayLoadMoreData()) {
                    presenter.nextPage();
                }
            }
        });

        swipeRefreshLayout.setOnRefreshListener(this);

        return root;
    }

    private boolean mayLoadMoreData() {
        return !themesRecyclerView.canScrollVertically(1) && isConnected;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getActivity().setTitle(getString(R.string.app_name));

        presenter.setView(this);

        if (baseActivity != null) {
            baseActivity.addOnConnectionListener(this);
        }

        //В случае если фрагмент не загрузил данные - загружаем первую страницу.
        // Пропускаем если данные уже есть и нам не нужно их заменять
        if (themes.size() == 0) {
            presenter.firstPage();
        }
    }


    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelableArrayList(Constants.THEME_MODEL, (ArrayList<? extends Parcelable>) themes);
    }

    @Override
    public void showThemes(List<Theme> themeList) {

        if (swipeRefreshLayout.isRefreshing()) {
            swipeRefreshLayout.setRefreshing(false);
        }
        themesProgress.setVisibility(View.GONE);

        if (themeList.size() == 0) {
            onError(getString(R.string.no_themes));
        }

        themes.clear();
        themes.addAll(themeList);
        themesListingAdapter.notifyDataSetChanged();
    }

    @Override
    public void onLoading() {
        themesEmptyView.setVisibility(View.GONE);

        if (!swipeRefreshLayout.isRefreshing()) {
            themesProgress.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onError(String errorMessage) {
//        themesRecyclerView.setVisibility(View.GONE);
        themesProgress.setVisibility(View.GONE);
        themesEmptyView.setText(errorMessage);
    }

    @Override
    public void onThemeClicked(Theme theme) {
        if (callback != null) {
            callback.onThemeClicked(theme);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (baseActivity != null) {
            baseActivity.removeOnConnectionListener(this);
        }
        presenter.destroy();
        unbinder.unbind();
    }

    @Override
    public void internetConnectionChanged(boolean connected) {
        isConnected = connected;
        if (mayLoadMoreData()) {
            presenter.nextPage();
        }
    }

    @Override
    public void onRefresh() {
        if (isConnected) {
            swipeRefreshLayout.setRefreshing(true);
            presenter.loadNewData();
        } else {
            swipeRefreshLayout.setRefreshing(false);
        }
    }

    @Override
    public void onDetach() {
        callback = null;
        baseActivity = null;
        super.onDetach();
    }

    public interface Callback {
        void onThemeClicked(Theme theme);
    }
}

File: app/src/main/java/com/example/sasham/testproject/themes/ThemesListingFragment.kt
package com.example.sasham.testproject.themes

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.arellomobile.mvp.presenter.InjectPresenter
import com.example.sasham.testproject.Constants
import com.example.sasham.testproject.MvpAppCompatFragment
import com.example.sasham.testproject.R
import com.example.sasham.testproject.base.BaseActivity
import com.example.sasham.testproject.model.Theme
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.fragment_themes_listing.*
import java.util.*


class ThemesListingFragment : MvpAppCompatFragment(), ThemesListingView, BaseActivity.OnConnectionListener, androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener {

    @InjectPresenter
    lateinit var presenter: ThemesListingPresenterImp

    private var callback: Callback? = null
    private val themes = ArrayList<Theme>()

    private var themesListingAdapter: ThemesListingAdapter? = null
    private var baseActivity: BaseActivity? = null
    private var isConnected = false

    override fun onAttach(context: Context) {
        super.onAttach(context)

        if (context is Callback) {
            callback = context
        }
        if (context is BaseActivity) {
            baseActivity = context
        }

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val root = inflater!!.inflate(R.layout.fragment_themes_listing, container, false)
        return root
    }

    private fun mayLoadMoreData(): Boolean {
        return !themesRecyclerView!!.canScrollVertically(1) && isConnected
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        themesListingAdapter = ThemesListingAdapter(themes, this)
        val layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        themesRecyclerView!!.adapter = themesListingAdapter
        themesRecyclerView!!.layoutManager = layoutManager

        //If we scrolled to end list then load more data
        themesRecyclerView!!.addOnScrollListener(object : androidx.recyclerview.widget.RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)

                if (mayLoadMoreData()) {
                    presenter!!.nextPage()
                }
            }
        })

        swipeRefreshLayout!!.setOnRefreshListener(this)

        activity!!.title = getString(R.string.app_name)

        if (baseActivity != null) {
            baseActivity!!.addOnConnectionListener(this)
        }

        //В случае если фрагмент не загрузил данные - загружаем первую страницу.
        // Пропускаем если данные уже есть и нам не нужно их заменять
        if (themes.size == 0) {
            presenter!!.firstPage()
        }
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState!!.putParcelableArrayList(Constants.THEME_MODEL, themes as ArrayList<out Parcelable>)
    }

    override fun showThemes(themeList: List<Theme>) {

        if (swipeRefreshLayout!!.isRefreshing) {
            swipeRefreshLayout!!.isRefreshing = false
        }
        themesProgress!!.visibility = View.GONE

        if (themeList.size == 0) {
            message(getString(R.string.no_themes))
        }

        themes.clear()
        themes.addAll(themeList)
        themesListingAdapter!!.notifyDataSetChanged()
    }

    override fun onLoading() {
        themesEmptyView!!.visibility = View.GONE

        if (!swipeRefreshLayout!!.isRefreshing) {
            themesProgress!!.visibility = View.VISIBLE
        }
    }

    override fun message(errorMessage: String?) {
        //        themesRecyclerView.setVisibility(View.GONE);
        themesProgress!!.visibility = View.GONE
        themesEmptyView!!.text = errorMessage
    }

    override fun onThemeClicked(theme: Theme) {
        if (callback != null) {
            callback!!.onThemeClicked(theme)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (baseActivity != null) {
            baseActivity!!.removeOnConnectionListener(this)
        }
    }

    override fun internetConnectionChanged(connected: Boolean) {
        isConnected = connected
        if (mayLoadMoreData()) {
            presenter!!.nextPage()
        }
    }

    override fun onRefresh() {
        if (isConnected) {
            swipeRefreshLayout!!.isRefreshing = true
            presenter!!.loadNewData()
        } else {
            swipeRefreshLayout!!.isRefreshing = false
        }
    }

    override fun onDetach() {
        callback = null
        baseActivity = null
        super.onDetach()
    }

    interface Callback {
        fun onThemeClicked(theme: Theme)
    }
}// Required empty public constructor
